src/dodal/devices/undulator.py

```diff
@@ -1,20 +1,50 @@
+import asyncio
 from enum import Enum
 
-from ophyd_async.core import ConfigSignal, StandardReadable, soft_signal_r_and_setter
+import numpy as np
+from bluesky.protocols import Movable
+from numpy import argmin, ndarray
+from ophyd_async.core import (
+    AsyncStatus,
+    ConfigSignal,
+    StandardReadable,
+    soft_signal_r_and_setter,
+)
 from ophyd_async.epics.motion import Motor
 from ophyd_async.epics.signal import epics_signal_r
 
+from dodal.log import LOGGER
+
+from .util.lookup_tables import energy_distance_table
+
+
+class AccessError(Exception):
+    pass
+
+
+# Enable to allow testing when the beamline is down, do not change in production!
+TEST_MODE = False
```

Comment:
Can we define 2 different implementations of MovableUndulator and switch between them in the dodal module, rather than having this global and the branching behaviour within this class' set method? I think it's clearer what the intention is.

How does TEST_MODE differ from backing everything with simulated signals? TEST_MODE isn't mentioned on the linked tickets, what're the requirements and where are they defined?

Reply:
I just copied the TEST_MODE over from the dcm_undulator. I don't have the historical knowledge how it came to be

Reply:
@DominicOram  is it ok to keep this TEST_MODE like that in the undulator.py?

Reply:
https://github.com/DiamondLightSource/dodal/issues/754 describes the usecase and tracks fixing it more nicely. Can we just leave the code as it is, with a link to this issue?

Reply:
splendid, keeping as-is

Reply:
Can we then get a link to the github issue as part of the comment @stan-dot?

Reply:
ah right, I forgot

Reply:
solved now